libs/crucible_proc_macros/src/lib.rs

```diff
@@ -321,3 +334,89 @@ pub fn crux_spec_for(args: TokenStream, input: TokenStream) -> TokenStream {
+/// their symbolic instances. In the case of enumerations, all
+/// variants can be returned. Unions are not supported.
+#[proc_macro_derive(Symbolic)]
+pub fn symbolic_derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = input.ident;
+
+    let body = match input.data {
+        Data::Struct(data_struct) => {
+            match &data_struct.fields {
+                Fields::Named(fields_named) => {
+                    let field_inits = fields_named.named.iter().map(|f| {
+                        let fname = &f.ident;
+                        quote! { #fname: crucible::Symbolic::symbolic(desc) }
+                    });
+                    quote! { Self { #( #field_inits, )* } }
+                }
+                Fields::Unnamed(fields_unnamed) => {
+                    let field_inits = fields_unnamed.unnamed.iter().map(|_| {
+                        quote! { crucible::Symbolic::symbolic(desc) }
+                    });
+                    quote! { Self( #( #field_inits, )* ) }
+                }
+                Fields::Unit => {
+                    quote! { Self }
+                }
+            }
+        }
+
+        Data::Enum(data_enum) => {
+            let variants: Vec<_> = data_enum.variants.iter().collect();
+            let arms = variants.iter().enumerate().map(|(i, v)| {
+                let vname = &v.ident;
+                match &v.fields {
+                    Fields::Named(fields_named) => {
+                        let field_inits = fields_named.named.iter().map(|f| {
+                            let fname = &f.ident;
+                            quote! { #fname: crucible::Symbolic::symbolic(desc) }
+                        });
+                        quote! { #i => Self::#vname { #( #field_inits, )* } }
+                    }
+                    Fields::Unnamed(fields_unnamed) => {
+                        let field_inits = fields_unnamed.unnamed.iter().map(|_| {
+                            quote! { crucible::Symbolic::symbolic(desc) }
+                        });
+                        quote! { #i => Self::#vname( #( #field_inits, )* ) }
+                    }
+                    Fields::Unit => {
+                        quote! { #i => Self::#vname }
+                    }
+                }
+            });
+
+            let n = variants.len();
+            quote! {
+                {
+                    let variant: usize = crucible::Symbolic::symbolic_where("variant", |&i| i < #n);
+                    match variant {
+                        #( #arms, )*
+                        _ => crucible::crucible_assume_unreachable!(),
+                    }
+                }
+            }
+        }
+
+        Data::Union(_) =>
+            panic!("Unions are not supported by derive(Symbolic)"),
+    };
+
+    let tokens = quote! {
+        #[cfg(crux)]
+        impl crucible::Symbolic for #name {
```

Comment:
Might be worth documenting that this derive doesn't support generics for now.  (Or else add support - IIRC there are some helpers in `syn` that make it pretty convenient.)